patches/server/0833-Friction-API.patch

```diff
@@ -0,0 +1,153 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Noah van der Aa <ndvdaa@gmail.com>
+Date: Wed, 15 Sep 2021 20:44:22 +0200
+Subject: [PATCH] Friction API
+
+
+diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
+index fe7e22d9a69d69dfcce63faa28e90945ea45fc49..8683161b8850cd5f7c806cba178b7792c109b5c9 100644
+--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
++++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
+@@ -260,6 +260,7 @@ public abstract class LivingEntity extends Entity {
+     public boolean bukkitPickUpLoot;
+     public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
+     public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
++    public io.papermc.paper.entity.Frictionable.FrictionState frictionState = io.papermc.paper.entity.Frictionable.FrictionState.UNSET; // Paper
+ 
+     @Override
+     public float getBukkitYaw() {
+@@ -702,7 +703,7 @@ public abstract class LivingEntity extends Entity {
+     }
+ 
+     public boolean shouldDiscardFriction() {
+-        return this.discardFriction;
++        return this.frictionState == io.papermc.paper.entity.Frictionable.FrictionState.IGNORE || (this.discardFriction && this.frictionState != io.papermc.paper.entity.Frictionable.FrictionState.ALWAYS); // Paper
+     }
+ 
+     public void setDiscardFriction(boolean noDrag) {
+@@ -726,6 +727,11 @@ public abstract class LivingEntity extends Entity {
+ 
+     @Override
+     public void addAdditionalSaveData(CompoundTag nbt) {
++        // Paper start
++        if (this.frictionState != io.papermc.paper.entity.Frictionable.FrictionState.UNSET) {
++            nbt.putString("paper.FrictionState", this.frictionState.toString());
++        }
++        // Paper end
+         nbt.putFloat("Health", this.getHealth());
+         nbt.putShort("HurtTime", (short) this.hurtTime);
+         nbt.putInt("HurtByTimestamp", this.lastHurtByMobTimestamp);
+@@ -768,6 +774,15 @@ public abstract class LivingEntity extends Entity {
+             absorptionAmount = 0;
+         }
+         this.setAbsorptionAmount(absorptionAmount);
++
++        if (nbt.contains("paper.FrictionState")) {
++            String fs = nbt.getString("paper.FrictionState");
++            try {
++                frictionState = io.papermc.paper.entity.Frictionable.FrictionState.valueOf(fs);
++            } catch (Exception ignored) {
++                org.apache.logging.log4j.LogManager.getLogger().error("Unknown FrictionState " + fs + " for " + this);
```

Comment:
The entity parent class already defines:
```java
protected static final Logger LOGGER = LogManager.getLogger();
```
which is used internally to log invalid data (see entities custom name parsing in `Entity#load(CompoundTag)` or zombie villagers loading their villager data.

This may just use that already initialized logger, as `getLogger` will have to walk the stack iirc.